api/src/services/security-service.ts

```diff
@@ -330,53 +330,94 @@ export class SecurityService extends DBService {
-   * If the override is set to `true` all security rules for the given set of features will be removed.
-   * Then the new security rules will be applied as normal.
+   * Patches security rules that are applied or removed to the given set of submission features. If a
+   * particular rule happens to belong to both `applyRuleIds` and `removeRuleIds`, it will always be
+   * added.
    *
-   * @param {number[]} features
-   * @param {number[]} rules
-   * @param {boolean}
-   * @return {*}  {Promise<SubmissionFeatureSecurityRecord[]>}
+   * @param {number[]} submissionFeatureIds IDs of the submission features whose security will be updated.
+   * @param {number[]} applyRuleIds IDs of the rules which will be applied after the patch operation
+   * @param {number[]} removeRuleIds IDs of the rules which will be removed after the patch operation
+   * @return {*}  {Promise<void>}
    * @memberof SecurityService
    */
-  async applySecurityRulesToSubmissionFeatures(
-    features: number[],
-    rules: number[],
-    override = false
-  ): Promise<SubmissionFeatureSecurityRecord[]> {
-    if (override) {
-      // we want to override any security rules present and can achieve this by remove everything first
-      await this.securityRepository.removeSecurityRulesFromSubmissionFeatures(features);
+  async patchSecurityRulesOnSubmissionFeatures(
+    submissionFeatureIds: number[],
+    applyRuleIds: number[],
+    removeRuleIds: number[]
+  ): Promise<void> {
+    defaultLog.debug({ label: 'patchSecurityRulesOnSubmissionFeatures', applyRuleIds, removeRuleIds });
+
+    if (!submissionFeatureIds.length) {
+      return;
     }
 
-    return this.securityRepository.applySecurityRulesToSubmissionFeatures(features, rules);
+    if (removeRuleIds.length > 0) {
+      await this.securityRepository.removeSecurityRulesFromSubmissionFeatures(submissionFeatureIds, removeRuleIds);
+    }
+
+    if (applyRuleIds.length > 0) {
+      await this.securityRepository.applySecurityRulesToSubmissionFeatures(submissionFeatureIds, applyRuleIds);
+    }
   }
 
   /**
-   * Removes all security rules for the given set of submission feature ids
+   * Removes the given security rules from the given set of submission feature ids. If
+   * no security rules ID is provided, all security rules will be removed for the given set
+   * of subission features.
+   *
+   * @param {number[]} submissionFeatureIds
+   * @param {number[]} [removeRuleIds]
    *
    * @return {*}  {Promise<SubmissionFeatureSecurityRecord[]>}
    * @memberof SecurityService
    */
-  async removeSecurityRulesFromSubmissionFeatures(features: number[]): Promise<SubmissionFeatureSecurityRecord[]> {
-    return this.securityRepository.removeSecurityRulesFromSubmissionFeatures(features);
+  async removeSecurityRulesFromSubmissionFeatures(
+    submissionFeatureIds: number[],
+    removeRuleIds?: number[]
+  ): Promise<SubmissionFeatureSecurityRecord[]> {
+    if (!submissionFeatureIds.length) {
```

Comment:
Is this check required when a similar check is already called in the parent function? redundancy is good just wondering if this function will be called from anywhere else.

Reply:
(Hoping this reply sufficiently addresses the other comments) I think the ethos behind service-level classes suggests that they may be called from both various endpoints and auxiliary services. For that reason, I added this check to all the new service-level methods, and removed it from the underlying repository methods.

To that end, I think the checks are worth it because they add virtually no overhead, and because they help to potentially avoid an unhelpful and costly operation (making a round trip to the database and returning an empty response).